// aws-cdk/lib/cxapp/cloud-executable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudExecutable = void 0;
const toolkit_lib_1 = require("@aws-cdk/toolkit-lib");
const cloud_assembly_1 = require("./cloud-assembly");
const api_private_1 = require("../../lib/api-private");
const singleton_plugin_host_1 = require("../cli/singleton-plugin-host");
const error_1 = require("../cli/telemetry/error");
const messages_1 = require("../cli/telemetry/messages");
const contextproviders = require("../context-providers");
/**
 * Represent the Cloud Executable and the synthesis we can do on it
 */
class CloudExecutable {
    constructor(props) {
        this.props = props;
    }
    async produce() {
        const synthesisResult = await this.synthesize(true);
        // We must return an `IReadableCloudAssembly` here, but this Cloud Assembly is only used in the context
        // of the CLI and `cli.ts` currently manages its own locking in the "synthesizer" callback function.
        //
        // All the lock-related functions are therefore no-ops.
        return new api_private_1.BorrowedAssembly(synthesisResult.assembly);
    }
    /**
     * Return whether there is an app command from the configuration
     */
    get hasApp() {
        return !!this.props.configuration.settings.get(['app']);
    }
    /**
     * Synthesize a set of stacks.
     *
     * @param cacheCloudAssembly - whether to cache the Cloud Assembly after it has been first synthesized.
     *   This is 'true' by default, and only set to 'false' for 'cdk watch',
     *   which needs to re-synthesize the Assembly each time it detects a change to the project files
     */
    async synthesize(cacheCloudAssembly = true) {
        if (!this._cloudAssembly || !cacheCloudAssembly) {
            this._cloudAssembly = await this.doSynthesize();
        }
        return this._cloudAssembly;
    }
    async doSynthesize() {
        // We may need to run the cloud executable multiple times in order to satisfy all missing context
        // (When the executable runs, it will tell us about context it wants to use
        // but it missing. We'll then look up the context and run the executable again, and
        // again, until it doesn't complain anymore or we've stopped making progress).
        let previouslyMissingKeys;
        const synthSpan = await this.props.ioHelper.span(messages_1.CLI_PRIVATE_SPAN.SYNTH_ASSEMBLY).begin({});
        let error;
        try {
            while (true) {
                const assembly = await this.props.synthesizer(this.props.sdkProvider, this.props.configuration);
                if (assembly.manifest.missing && assembly.manifest.missing.length > 0) {
                    const missingKeys = missingContextKeys(assembly.manifest.missing);
                    if (!this.canLookup) {
                        throw new toolkit_lib_1.ToolkitError('Context lookups have been disabled. '
                            + 'Make sure all necessary context is already in \'cdk.context.json\' by running \'cdk synth\' on a machine with sufficient AWS credentials and committing the result. '
                            + `Missing context keys: '${Array.from(missingKeys).join(', ')}'`);
                    }
                    let tryLookup = true;
                    if (previouslyMissingKeys && setsEqual(missingKeys, previouslyMissingKeys)) {
                        await this.props.ioHelper.defaults.debug('Not making progress trying to resolve environmental context. Giving up.');
                        tryLookup = false;
                    }
                    previouslyMissingKeys = missingKeys;
                    if (tryLookup) {
                        await this.props.ioHelper.defaults.debug('Some context information is missing. Fetching...');
                        const updates = await contextproviders.provideContextValues(assembly.manifest.missing, this.props.sdkProvider, singleton_plugin_host_1.GLOBAL_PLUGIN_HOST, this.props.ioHelper);
                        for (const [key, value] of Object.entries(updates)) {
                            this.props.configuration.context.set(key, value);
                        }
                        // Cache the new context to disk
                        await this.props.configuration.saveContext();
                        // Execute again
                        continue;
                    }
                }
                return new cloud_assembly_1.CloudAssembly(assembly, this.props.ioHelper);
            }
        }
        catch (e) {
            error = {
                name: (0, error_1.cdkCliErrorName)(e.name),
            };
            throw e;
        }
        finally {
            await synthSpan.end({ error });
        }
    }
    get canLookup() {
        return !!(this.props.configuration.settings.get(['lookups']) ?? true);
    }
}
exports.CloudExecutable = CloudExecutable;
/**
 * Return all keys of missing context items
 */
function missingContextKeys(missing) {
    return new Set((missing || []).map(m => m.key));
}
function setsEqual(a, b) {
    if (a.size !== b.size) {
        return false;
    }
    for (const x of a) {
        if (!b.has(x)) {
            return false;
        }
    }
    return true;
}
