// aws-cdk/lib/cli/proxy-agent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxyAgentProvider = void 0;
const fs = require("fs-extra");
const proxy_agent_1 = require("proxy-agent");
class ProxyAgentProvider {
    constructor(ioHelper) {
        this.ioHelper = ioHelper;
    }
    async create(options) {
        // Force it to use the proxy provided through the command line.
        // Otherwise, let the ProxyAgent auto-detect the proxy using environment variables.
        const getProxyForUrl = options.proxyAddress != null
            ? () => Promise.resolve(options.proxyAddress)
            : undefined;
        return new proxy_agent_1.ProxyAgent({
            ca: await this.tryGetCACert(options.caBundlePath),
            getProxyForUrl,
        });
    }
    async tryGetCACert(bundlePath) {
        const path = bundlePath || this.caBundlePathFromEnvironment();
        if (path) {
            await this.ioHelper.defaults.debug(`Using CA bundle path: ${path}`);
            try {
                if (!fs.pathExistsSync(path)) {
                    return undefined;
                }
                return fs.readFileSync(path, { encoding: 'utf-8' });
            }
            catch (e) {
                await this.ioHelper.defaults.debug(String(e));
                return undefined;
            }
        }
        return undefined;
    }
    /**
     * Find and return a CA certificate bundle path to be passed into the SDK.
     */
    caBundlePathFromEnvironment() {
        if (process.env.aws_ca_bundle) {
            return process.env.aws_ca_bundle;
        }
        if (process.env.AWS_CA_BUNDLE) {
            return process.env.AWS_CA_BUNDLE;
        }
        return undefined;
    }
}
exports.ProxyAgentProvider = ProxyAgentProvider;
