// aws-cdk/lib/commands/flags/interactive-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractiveHandler = void 0;
// @ts-ignore
const enquirer_1 = require("enquirer");
const types_1 = require("./types");
class InteractiveHandler {
    constructor(flags, flagOperations) {
        this.flags = flags;
        this.flagOperations = flagOperations;
    }
    /** Displays flags that have differences between user and recommended values */
    async displayFlagsWithDifferences() {
        const flagsWithDifferences = this.flags.filter(flag => flag.userValue === undefined || !this.isUserValueEqualToRecommended(flag));
        if (flagsWithDifferences.length > 0) {
            await this.flagOperations.displayFlagTable(flagsWithDifferences);
        }
    }
    /** Checks if user value matches recommended value */
    isUserValueEqualToRecommended(flag) {
        return String(flag.userValue) === String(flag.recommendedValue);
    }
    /** Main interactive mode handler that shows menu and processes user selection */
    async handleInteractiveMode() {
        await this.displayFlagsWithDifferences();
        const prompt = new enquirer_1.Select({
            name: 'option',
            message: 'Menu',
            choices: Object.values(types_1.FlagsMenuOptions),
        });
        const answer = await prompt.run();
        switch (answer) {
            case types_1.FlagsMenuOptions.ALL_TO_RECOMMENDED:
                return { recommended: true, all: true, set: true };
            case types_1.FlagsMenuOptions.UNCONFIGURED_TO_RECOMMENDED:
                return { recommended: true, unconfigured: true, set: true };
            case types_1.FlagsMenuOptions.UNCONFIGURED_TO_DEFAULT:
                return { default: true, unconfigured: true, set: true };
            case types_1.FlagsMenuOptions.MODIFY_SPECIFIC_FLAG:
                return this.handleSpecificFlagSelection();
            case types_1.FlagsMenuOptions.EXIT:
                return null;
            default:
                return null;
        }
    }
    /** Handles the specific flag selection flow with flag and value prompts */
    async handleSpecificFlagSelection() {
        const booleanFlags = this.flags.filter(flag => this.flagOperations.isBooleanFlag(flag));
        const flagPrompt = new enquirer_1.Select({
            name: 'flag',
            message: 'Select which flag you would like to modify:',
            limit: 100,
            choices: booleanFlags.map(flag => flag.name),
        });
        const selectedFlagName = await flagPrompt.run();
        const valuePrompt = new enquirer_1.Select({
            name: 'value',
            message: 'Select a value:',
            choices: ['true', 'false'],
        });
        const value = await valuePrompt.run();
        return {
            FLAGNAME: [selectedFlagName],
            value,
            set: true,
        };
    }
}
exports.InteractiveHandler = InteractiveHandler;
