// aws-cdk/lib/commands/flags/router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlagOperationRouter = void 0;
class FlagOperationRouter {
    constructor(validator, interactiveHandler, flagOperations) {
        this.validator = validator;
        this.interactiveHandler = interactiveHandler;
        this.flagOperations = flagOperations;
    }
    /** Routes flag operations to appropriate handlers based on parameters */
    async route(params) {
        if (params.interactive) {
            await this.handleInteractiveMode();
            return;
        }
        if (params.safe) {
            await this.flagOperations.setSafeFlags(params);
            return;
        }
        const isValid = await this.validator.validateParams(params);
        if (!isValid)
            return;
        if (params.set) {
            await this.handleSetOperations(params);
        }
        else {
            await this.flagOperations.displayFlags(params);
            await this.showHelpMessage(params);
        }
    }
    /** Handles flag setting operations, routing to single or multiple flag methods */
    async handleSetOperations(params) {
        if (params.FLAGNAME && params.value) {
            await this.flagOperations.setFlag(params);
        }
        else if (params.all || params.unconfigured) {
            await this.flagOperations.setMultipleFlags(params);
        }
    }
    /** Manages interactive mode */
    async handleInteractiveMode() {
        while (true) {
            const interactiveParams = await this.interactiveHandler.handleInteractiveMode();
            if (!interactiveParams)
                return;
            await this.flagOperations.execute(interactiveParams);
            if (!interactiveParams.FLAGNAME) {
                return;
            }
        }
    }
    /** Shows help message when no specific options are provided */
    async showHelpMessage(params) {
        if (!params.all && !params.FLAGNAME) {
            await this.flagOperations.displayHelpMessage();
        }
    }
}
exports.FlagOperationRouter = FlagOperationRouter;
