// aws-cdk/lib/legacy/index.js
"use strict";
// This is a barrel export file, of all known symbols that are imported by users from the `aws-cdk` package.
// Importing these symbols was never officially supported, but here we are.
// In order to preserver backwards-compatibly for these users, we re-export and preserve them as explicit subpath exports.
// See https://github.com/aws/aws-cdk/pull/33021 for more information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.lowerCaseFirstCharacter = exports.deepMerge = exports.contentHash = exports.leftPad = exports.partition = exports.numberFromBool = exports.isEmpty = exports.isArray = exports.ifDefined = exports.flatten = exports.deepClone = exports.describe = exports.command = exports.aliases = exports.availableInitTemplates = exports.RequireApproval = exports.versionNumber = exports.formatAsBanner = exports.rootDir = exports.exec = exports.cli = exports.execProgram = exports.CloudExecutable = exports.Bootstrapper = exports.Settings = exports.PluginHost = exports.deployStack = exports.Deployments = exports.CfnEvaluationException = exports.PROJECT_CONTEXT = exports.Configuration = exports.Command = void 0;
// Legacy Code Copies
__exportStar(require("./logging"), exports);
__exportStar(require("./aws-auth"), exports);
var configuration_1 = require("./configuration");
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return configuration_1.Command; } });
Object.defineProperty(exports, "Configuration", { enumerable: true, get: function () { return configuration_1.Configuration; } });
Object.defineProperty(exports, "PROJECT_CONTEXT", { enumerable: true, get: function () { return configuration_1.PROJECT_CONTEXT; } });
// API
var cloudformation_1 = require("../api/cloudformation");
Object.defineProperty(exports, "CfnEvaluationException", { enumerable: true, get: function () { return cloudformation_1.CfnEvaluationException; } });
var deployments_1 = require("../api/deployments");
Object.defineProperty(exports, "Deployments", { enumerable: true, get: function () { return deployments_1.Deployments; } });
var api_private_1 = require("../api-private");
Object.defineProperty(exports, "deployStack", { enumerable: true, get: function () { return api_private_1.deployStack; } });
var plugin_1 = require("../api/plugin");
Object.defineProperty(exports, "PluginHost", { enumerable: true, get: function () { return plugin_1.PluginHost; } });
var settings_1 = require("../api/settings");
Object.defineProperty(exports, "Settings", { enumerable: true, get: function () { return settings_1.Settings; } });
var bootstrap_1 = require("../api/bootstrap");
Object.defineProperty(exports, "Bootstrapper", { enumerable: true, get: function () { return bootstrap_1.Bootstrapper; } });
// CLI
var cxapp_1 = require("../cxapp");
Object.defineProperty(exports, "CloudExecutable", { enumerable: true, get: function () { return cxapp_1.CloudExecutable; } });
Object.defineProperty(exports, "execProgram", { enumerable: true, get: function () { return cxapp_1.execProgram; } });
var cli_1 = require("../cli/cli");
Object.defineProperty(exports, "cli", { enumerable: true, get: function () { return cli_1.cli; } });
Object.defineProperty(exports, "exec", { enumerable: true, get: function () { return cli_1.exec; } });
var root_dir_1 = require("../cli/root-dir");
Object.defineProperty(exports, "rootDir", { enumerable: true, get: function () { return root_dir_1.cliRootDir; } });
var console_formatters_1 = require("../cli/util/console-formatters");
Object.defineProperty(exports, "formatAsBanner", { enumerable: true, get: function () { return console_formatters_1.formatAsBanner; } });
var version_1 = require("../cli/version");
Object.defineProperty(exports, "versionNumber", { enumerable: true, get: function () { return version_1.versionNumber; } });
// Commands
var cloud_assembly_schema_1 = require("@aws-cdk/cloud-assembly-schema");
Object.defineProperty(exports, "RequireApproval", { enumerable: true, get: function () { return cloud_assembly_schema_1.RequireApproval; } });
var init_1 = require("../commands/init");
Object.defineProperty(exports, "availableInitTemplates", { enumerable: true, get: function () { return init_1.availableInitTemplates; } });
var docs_1 = require("../commands/docs");
Object.defineProperty(exports, "aliases", { enumerable: true, get: function () { return docs_1.aliases; } });
Object.defineProperty(exports, "command", { enumerable: true, get: function () { return docs_1.command; } });
Object.defineProperty(exports, "describe", { enumerable: true, get: function () { return docs_1.describe; } });
// util
var util_1 = require("../util");
Object.defineProperty(exports, "deepClone", { enumerable: true, get: function () { return util_1.deepClone; } });
Object.defineProperty(exports, "flatten", { enumerable: true, get: function () { return util_1.flatten; } });
Object.defineProperty(exports, "ifDefined", { enumerable: true, get: function () { return util_1.ifDefined; } });
Object.defineProperty(exports, "isArray", { enumerable: true, get: function () { return util_1.isArray; } });
Object.defineProperty(exports, "isEmpty", { enumerable: true, get: function () { return util_1.isEmpty; } });
Object.defineProperty(exports, "numberFromBool", { enumerable: true, get: function () { return util_1.numberFromBool; } });
Object.defineProperty(exports, "partition", { enumerable: true, get: function () { return util_1.partition; } });
Object.defineProperty(exports, "leftPad", { enumerable: true, get: function () { return util_1.padLeft; } });
Object.defineProperty(exports, "contentHash", { enumerable: true, get: function () { return util_1.contentHash; } });
Object.defineProperty(exports, "deepMerge", { enumerable: true, get: function () { return util_1.deepMerge; } });
Object.defineProperty(exports, "lowerCaseFirstCharacter", { enumerable: true, get: function () { return util_1.lowerCaseFirstCharacter; } });
