// aws-cdk/lib/cli/telemetry/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLI_TELEMETRY_CODES = exports.CLI_PRIVATE_SPAN = exports.CLI_PRIVATE_IO = void 0;
const make = require("../../api-private");
/**
 * Private message types specific to the CLI
 */
exports.CLI_PRIVATE_IO = {
    CDK_CLI_I1000: make.trace({
        code: 'CDK_CLI_I1000',
        description: 'Cloud Execution is starting',
        interface: 'EventStart',
    }),
    CDK_CLI_I1001: make.trace({
        code: 'CDK_CLI_I1001',
        description: 'Cloud Executable Result',
        interface: 'EventResult',
    }),
    CDK_CLI_I2000: make.trace({
        code: 'CDK_CLI_I2000',
        description: 'Command has started',
        interface: 'EventStart',
    }),
    CDK_CLI_I2001: make.trace({
        code: 'CDK_CLI_I2001',
        description: 'Command has finished executing',
        interface: 'EventResult',
    }),
    CDK_CLI_I3000: make.trace({
        code: 'CDK_CLI_I3000',
        description: 'Deploy has started',
        interface: 'EventStart',
    }),
    CDK_CLI_I3001: make.trace({
        code: 'CDK_CLI_I3001',
        description: 'Deploy has finished',
        interface: 'EventResult',
    }),
};
/**
 * Payload type of the end message must extend Duration
 */
exports.CLI_PRIVATE_SPAN = {
    SYNTH_ASSEMBLY: {
        name: 'Synthesis',
        start: exports.CLI_PRIVATE_IO.CDK_CLI_I1000,
        end: exports.CLI_PRIVATE_IO.CDK_CLI_I1001,
    },
    COMMAND: {
        name: 'Command',
        start: exports.CLI_PRIVATE_IO.CDK_CLI_I2000,
        end: exports.CLI_PRIVATE_IO.CDK_CLI_I2001,
    },
    DEPLOY: {
        name: 'Deploy',
        start: exports.CLI_PRIVATE_IO.CDK_CLI_I3000,
        end: exports.CLI_PRIVATE_IO.CDK_CLI_I3001,
    },
};
exports.CLI_TELEMETRY_CODES = [
    exports.CLI_PRIVATE_IO.CDK_CLI_I1001,
    exports.CLI_PRIVATE_IO.CDK_CLI_I2001,
    exports.CLI_PRIVATE_IO.CDK_CLI_I3001,
];
