// aws-cdk/lib/cli/telemetry/sanitation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sanitizeCommandLineArguments = sanitizeCommandLineArguments;
exports.sanitizeContext = sanitizeContext;
const feature_flags_1 = require("./feature-flags");
/**
 * argv is the output of yargs
 */
function sanitizeCommandLineArguments(argv) {
    // Get the configuration of the arguments
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const config = require('../cli-type-registry.json');
    const command = argv._[0];
    const path = [command];
    const parameters = {};
    const globalOptions = Object.keys(config.globalOptions);
    const commandOptions = Object.keys(config.commands[command]?.options ?? {});
    const commandArg = config.commands[command]?.arg;
    for (const argName of Object.keys(argv)) {
        if (argName === commandArg?.name) {
            if (commandArg.variadic) {
                for (let i = 0; i < argv[argName].length; i++) {
                    path.push(`$${argName}_${i + 1}`);
                }
            }
            else {
                path.push(`$${argName}`);
            }
        }
        // Continue if the arg name is not a global option or command option
        // arg name comes from yargs and could be an alias; we trust that the "normal"
        // name has the same information and that is what we want to record
        if (argv[argName] === undefined || (!globalOptions.includes(argName) && !commandOptions.includes(argName))) {
            continue;
        }
        if (isNumberOrBoolean(argv[argName]) || isKnownEnumValue(argName, argv[argName], command, config)) {
            parameters[argName] = argv[argName];
        }
        else {
            parameters[argName] = '<redacted>';
        }
    }
    return {
        path,
        parameters,
    };
}
function sanitizeContext(context) {
    const sanitizedContext = {};
    for (const [flag, value] of Object.entries(context.all)) {
        // Skip if flag is not in the FeatureFlags enum
        if (!isFeatureFlag(flag)) {
            continue;
        }
        // Falsy options include boolean false, string 'false'
        // All other inputs evaluate to true
        const sanitizedValue = isBoolean(value) ? value : (value !== 'false');
        sanitizedContext[flag] = sanitizedValue;
    }
    return sanitizedContext;
}
function isBoolean(value) {
    return typeof value === 'boolean';
}
function isNumberOrBoolean(value) {
    return typeof value === 'number' || isBoolean(value);
}
function isKnownEnumValue(name, value, command, config) {
    const propertyDefiniton = config.globalOptions[name] ?? config.commands[command]?.options[name];
    if (propertyDefiniton.type === 'string') {
        // Even if the property has choices, only record if the value is a valid choice
        return propertyDefiniton.choices?.includes(value);
    }
    return false;
}
function isFeatureFlag(flag) {
    return Object.values(feature_flags_1.FeatureFlag).includes(flag);
}
