// aws-cdk/lib/cli/telemetry/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelemetrySession = void 0;
const crypto_1 = require("crypto");
const toolkit_lib_1 = require("@aws-cdk/toolkit-lib");
const installation_id_1 = require("./installation-id");
const library_version_1 = require("./library-version");
const sanitation_1 = require("./sanitation");
const schema_1 = require("./schema");
const ci_systems_1 = require("../ci-systems");
const messages_1 = require("../telemetry/messages");
const ci_1 = require("../util/ci");
const version_1 = require("../version");
const ABORTED_ERROR_MESSAGE = '__CDK-Toolkit__Aborted';
class TelemetrySession {
    constructor(props) {
        this.props = props;
        this.count = 0;
        this.ioHost = props.ioHost;
        this.client = props.client;
    }
    async begin() {
        // sanitize the raw cli input
        const { path, parameters } = (0, sanitation_1.sanitizeCommandLineArguments)(this.props.arguments);
        this._sessionInfo = {
            identifiers: {
                installationId: await (0, installation_id_1.getOrCreateInstallationId)(this.ioHost.asIoHelper()),
                sessionId: (0, crypto_1.randomUUID)(),
                telemetryVersion: '1.0',
                cdkCliVersion: (0, version_1.versionNumber)(),
                cdkLibraryVersion: await (0, library_version_1.getLibraryVersion)(this.ioHost.asIoHelper()),
            },
            event: {
                command: {
                    path,
                    parameters,
                    config: {
                        context: (0, sanitation_1.sanitizeContext)(this.props.context),
                    },
                },
            },
            environment: {
                ci: (0, ci_1.isCI)() || Boolean((0, ci_systems_1.detectCiSystem)()),
                os: {
                    platform: process.platform,
                    release: process.release.name,
                },
                nodeVersion: process.version,
            },
            project: {},
        };
        // If SIGINT has a listener installed, its default behavior will be removed (Node.js will no longer exit).
        // This ensures that on SIGINT we process safely close the telemetry session before exiting.
        process.on('SIGINT', async () => {
            try {
                await this.end({
                    name: schema_1.ErrorName.TOOLKIT_ERROR,
                    message: ABORTED_ERROR_MESSAGE,
                });
            }
            catch (e) {
                await this.ioHost.defaults.trace(`Ending Telemetry failed: ${e.message}`);
            }
            process.exit(1);
        });
        // Begin the session span
        this.span = await this.ioHost.asIoHelper().span(messages_1.CLI_PRIVATE_SPAN.COMMAND).begin({});
    }
    async attachRegion(region) {
        this.sessionInfo.identifiers = {
            ...this.sessionInfo.identifiers,
            region,
        };
    }
    /**
     * When the command is complete, so is the CliIoHost. Ends the span of the entire CliIoHost
     * and notifies with an optional error message in the data.
     */
    async end(error) {
        await this.span?.end({ error });
        // Ideally span.end() should no-op if called twice, but that is not the case right now
        this.span = undefined;
        await this.client.flush();
    }
    async emit(event) {
        this.count += 1;
        return this.client.emit({
            event: {
                command: this.sessionInfo.event.command,
                state: getState(event.error),
                eventType: event.eventType,
            },
            identifiers: {
                ...this.sessionInfo.identifiers,
                eventId: `${this.sessionInfo.identifiers.sessionId}:${this.count}`,
                timestamp: new Date().toISOString(),
            },
            environment: this.sessionInfo.environment,
            project: this.sessionInfo.project,
            duration: {
                total: event.duration,
            },
            ...(event.error ? {
                error: {
                    name: event.error.name,
                },
            } : {}),
        });
    }
    get sessionInfo() {
        if (!this._sessionInfo) {
            throw new toolkit_lib_1.ToolkitError('Session Info not initialized. Call begin() first.');
        }
        return this._sessionInfo;
    }
}
exports.TelemetrySession = TelemetrySession;
function getState(error) {
    if (error) {
        return isAbortedError(error) ? 'ABORTED' : 'FAILED';
    }
    return 'SUCCEEDED';
}
function isAbortedError(error) {
    if (error?.name === 'ToolkitError' && error?.message?.includes(ABORTED_ERROR_MESSAGE)) {
        return true;
    }
    return false;
}
